Fixes #4504

greglandrum opened the pull request:
The fix here is quite simple, but it's a bit buried in a *bunch* of code modernization.
I normally wouldn't do this, but the depiction code hadn't really been touched in years and I kept getting hung up on the C++98-ishness of it while trying to find the bug. Looking through all those code changes is probably a waste of time, but that's your call. :-)

The code change which fixes the bug is here:
https://github.com/greglandrum/rdkit/commit/9d53c37c2cf1e2efd46debd9045b7f3a185f2b7d#diff-6b894e21d231fd69d75e7eedd6607ede20c5698848d915d7891aef660e1f3288R705

The problem occurred when adding a fourth attachment point to an atom where the other 3 attachment points form a T: the resulting 180 degree angle screws up the code which detects where the next ligand should go. This fixes the problem by detecting the situation, trying both possibilities, and placing the ligand on the side where it causes the least collisions.

The other non-modernization change here:
The code wasn't really being particularly well tested: the majority of the tests just check to see that a conformer is created, not that the conformer is right. This commit converts one of those "tests" to actually check that the coordinates haven't changed: https://github.com/greglandrum/rdkit/commit/9098a65ef3527bf270844e7d960b380a33b7d10b#diff-6266054369a94106f70cf2cf98e33576c05a8457cf6323a6b690e32e71c89f58
Hopefully this doesn't break across platforms... the CI tests will show that.

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -292,12 +277,11 @@ EmbeddedFrag::EmbeddedFrag(const RDKit::Bond *dblBond) {
-  for (efi = d_eatoms.begin(); efi != d_eatoms.end(); efi++) {
-    RDGeom::Point2D rloc = efi->second.loc;
+  for (const auto &efi : d_eatoms) {
+    RDGeom::Point2D rloc = efi.second.loc;
```
This doesn't need to be a non-const copy, although I expect it really doesn't matter.

greglandrum replied:
good catch

bp-kelley replied:
I think these are going to end up optimized into register anyway...

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -350,10 +333,9 @@ void EmbeddedFrag::setupNewNeighs() {  // const RDKit::ROMol *mol) {
-  for (eci = d_eatoms.begin(); eci != d_eatoms.end(); eci++) {
-    unsigned int aid = eci->first;
+  for (const auto &eci : d_eatoms) {
+    unsigned int aid = eci.first;
```
You've been consolidating to auto in these cases, might as well be consistent.

greglandrum replied:
yep. will do

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -737,6 +691,17 @@ void EmbeddedFrag::addAtomToAtomWithAng(unsigned int aid, unsigned int toAid) {
   rtrans.SetTransform(refLoc, currAngle);
   RDGeom::Point2D currLoc = nb2;
   rtrans.TransformPoint(currLoc);
+  if (fabs(remAngle) - M_PI < 1e-3) {
```
Is this clash detection?

greglandrum replied:
This detects the condition which could lead to the clashes.
The clashes themselves are caught in the calls to `findNumNeigh()` a couple lines down

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -982,62 +940,58 @@ void EmbeddedFrag::mergeWithCommon(EmbeddedFrag &embObj,
   while (1) {
     RDKit::INT_VECT commAtms;
-    for (efri = efrags.begin(); efri != efrags.end(); efri++) {
+    for (auto efri = efrags.begin(); efri != efrags.end(); ++efri) {
```
any reason not to do
```
for( const auto &efri : efrags)
```

here?

greglandrum replied:
yeah, we need the iterator to set `nfri` a couple lines down

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1070,44 +1023,41 @@ void EmbeddedFrag::expandEfrag(RDKit::INT_LIST &nratms,
-        nfri = efrags.end();
-        for (efri = efrags.begin(); efri != efrags.end(); efri++) {
+        auto nfri = efrags.end();
+        for (auto efri = efrags.begin(); efri != efrags.end(); ++efri) {
```
Might as well convert to range, you're right @greglandrum, these are a bit of an eyesore now!

greglandrum replied:
same story, need the iterator later

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1296,30 +1231,31 @@ double EmbeddedFrag::mimicDistMatAndDensityCostFunc(
+    return 0;
+  }
+  auto dsize = na * (na - 1) / 2;
   auto *ddata2D = new double[dsize];
```
Switch to unique_ptr?

greglandrum replied:
it's added to a smart pointer on the next line

bp-kelley commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1346,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
   }
 
-  nt = nb + nd4;
+  unsigned int nt = nb + nd4;
```
auto?

greglandrum replied:
tried that: ends up being a size_t and producing warnings later.

bp-kelley commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -125,7 +121,7 @@ std::list<EmbeddedFrag>::iterator _findLargestFrag(
   std::list<EmbeddedFrag>::iterator mfri;
   int msiz = 0;
-  for (auto efri = efrags.begin(); efri != efrags.end(); efri++) {
+  for (auto efri = efrags.begin(); efri != efrags.end(); ++efri) {
```
Possibly use range

greglandrum replied:
same story with using the iterator elsewhere

bp-kelley commented on Code/GraphMol/Depictor/Wrap/testDepictor.py:
```diff
@@ -54,6 +54,8 @@ def compareCoords(m, molFile):
         pos = ci.GetAtomPosition(i)
         opos = co.GetAtomPosition(i)
         if not ptEq(pos, opos):
+            print(Chem.MolToMolBlock(m))
```
Perhaps also a string indicating what the error is here?

greglandrum replied:
I thought about that, but it's likely that there are going to be a lot of mismatched atom positions when this happens and this would only show the first. I'm 99% sure that you're going to have to look at the structure anyway in order to figure out what's going on, so it just didn't seem worth it.
I can add it if you really think it'll be helpful

bp-kelley replied:
I agree it probably won't be too useful, but I like a bit of verbiage when reporting test errors.

bp-kelley commented on Code/GraphMol/Depictor/testDepictor.cpp:
```diff
@@ -33,6 +33,29 @@ typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
+    pt2 -= pt1;
+
+    if ((fabs(pt2.x) >= tol) || (fabs(pt2.y) >= tol)) {
+      std::cerr << MolToMolBlock(*mol1, cid1) << std::endl;
```
Perhaps a string indicated what error occurred?

greglandrum replied:
same comment here

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
-        bi = rotBonds[ri];
-        this->flipAboutBond(bi);
-
+        this->flipAboutBond(rotBonds[ri]);
```
```suggestion
        this->flipAboutBond(rotBonds.at(ri));
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
         unsigned int d4i =
             ri - nb;  // so we will permute at the 'di'th degree 4 node
-        ai = deg4nodes[d4i];
+        auto ai = deg4nodes[d4i];
```
```suggestion
        auto ai = deg4nodes.at(d4i);
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
-             aci != deg4NbrAids[d4i].end(); aci++) {
-          nbrLocs.push_back(&(d_eatoms[*aci].loc));
+        for (auto aci : deg4NbrAids[d4i]) {
+          nbrLocs.push_back(&(d_eatoms[aci].loc));
```
```suggestion
          nbrLocs.push_back(&(d_eatoms.at(aci).loc));
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
         }
-        INT_PAIR_VECT bndPairs = findBondsPairsToPermuteDeg4(
-            d_eatoms[ai].loc, deg4NbrBids[d4i], nbrLocs);
+        auto bndPairs = findBondsPairsToPermuteDeg4(d_eatoms[ai].loc,
+                                                    deg4NbrBids[d4i], nbrLocs);
```
```suggestion
        auto bndPairs = findBondsPairsToPermuteDeg4(d_eatoms.at(ai).loc,
                                                    deg4NbrBids.at(d4i), nbrLocs);
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
-        aid1 = dp_mol->getBondWithIdx(bndPairs[fbi].first)->getOtherAtomIdx(ai);
-        aid2 =
+        auto aid1 =
+            dp_mol->getBondWithIdx(bndPairs[fbi].first)->getOtherAtomIdx(ai);
```
```suggestion
            dp_mol->getBondWithIdx(bndPairs.at(fbi).first)->getOtherAtomIdx(ai);
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
+        auto aid1 =
+            dp_mol->getBondWithIdx(bndPairs[fbi].first)->getOtherAtomIdx(ai);
+        auto aid2 =
             dp_mol->getBondWithIdx(bndPairs[fbi].second)->getOtherAtomIdx(ai);
```
```suggestion
            dp_mol->getBondWithIdx(bndPairs.at(fbi).second)->getOtherAtomIdx(ai);
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1423,69 +1345,68 @@ void EmbeddedFrag::randomSampleFlipsAndPermutations(
-  for (efi = d_eatoms.begin(); efi != d_eatoms.end(); efi++) {
-    efi->second.loc = bestCrdMap[efi->first];
+  for (auto &efi : d_eatoms) {
+    efi.second.loc = bestCrdMap[efi.first];
```
```suggestion
    efi.second.loc = bestCrdMap.at(efi.first);
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1593,96 +1511,85 @@ void _recurseDegTwoRingAtoms(unsigned int aid, const RDKit::ROMol *mol,
 
-  RDGeom::Point2D begLoc = d_eatoms[begAid].loc;
-  RDGeom::Point2D endLoc = d_eatoms[endAid].loc;
+  const auto &begLoc = d_eatoms[begAid].loc;
+  const auto &endLoc = d_eatoms[endAid].loc;
```
```suggestion
  const auto &begLoc = d_eatoms.at(begAid).loc;
  const auto &endLoc = d_eatoms.at(endAid).loc;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1769,16 +1663,15 @@ void EmbeddedFrag::openAngles(const double *dmat, unsigned int aid1,
-  bool fixed2 = d_eatoms[aid2].df_fixed;
+  auto deg1 = getDepictDegree(dp_mol->getAtomWithIdx(aid1));
+  auto deg2 = getDepictDegree(dp_mol->getAtomWithIdx(aid2));
+  auto fixed1 = d_eatoms[aid1].df_fixed;
+  auto fixed2 = d_eatoms[aid2].df_fixed;
```
```suggestion
  auto fixed1 = d_eatoms.at(aid1).df_fixed;
  auto fixed2 = d_eatoms.at(aid2).df_fixed;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1794,12 +1687,9 @@ void EmbeddedFrag::openAngles(const double *dmat, unsigned int aid1,
-  RDGeom::Point2D v2 = d_eatoms[aid1].loc - d_eatoms[aidA].loc;
-  RDGeom::Point2D v1 = d_eatoms[aidB].loc - d_eatoms[aidA].loc;
-  double cross = (v1.x) * (v2.y) - (v1.y) * (v2.x);
+  auto v2 = d_eatoms[aid1].loc - d_eatoms[aidA].loc;
+  auto v1 = d_eatoms[aidB].loc - d_eatoms[aidA].loc;
```
```suggestion
  auto v2 = d_eatoms.at(aid1).loc - d_eatoms.at(aidA).loc;
  auto v1 = d_eatoms.at(aidB).loc - d_eatoms.at(aidA).loc;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1832,96 +1722,78 @@ void EmbeddedFrag::openAngles(const double *dmat, unsigned int aid1,
+      auto rotBonds = getRotatableBonds(*dp_mol, cAids.first, cAids.second);
+      auto prevDensity = this->totalDensity();
+      for (auto ri : rotBonds) {
+        if ((doneBonds.find(ri) == doneBonds.end()) ||
+            (doneBonds[ri] < NUM_BONDS_FLIPS)) {
+          if (doneBonds.find(ri) == doneBonds.end()) {
+            doneBonds[ri] = 1;
```
```suggestion
        auto doneBondsRiIt = doneBonds.find(ri);
        if ((doneBondsRiIt == doneBonds.end()) ||
            (doneBondsRiIt->second < NUM_BONDS_FLIPS)) {
          if (doneBondsRiIt == doneBonds.end()) {
            doneBonds[ri] = 1;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1832,96 +1722,78 @@ void EmbeddedFrag::openAngles(const double *dmat, unsigned int aid1,
+            doneBonds[ri] = 1;
           } else {
-            doneBonds[*ri] += 1;
+            doneBonds[ri] += 1;
```
```suggestion
            doneBondsRiIt->second += 1;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1931,17 +1803,17 @@ void EmbeddedFrag::removeCollisionsShortenBonds() {
-    bool fixed2 = d_eatoms[aid2].df_fixed;
+    auto aid1 = cAids.first;
+    auto aid2 = cAids.second;
+    auto fixed1 = d_eatoms[aid1].df_fixed;
+    auto fixed2 = d_eatoms[aid2].df_fixed;
```
```suggestion
    auto fixed1 = d_eatoms.at(aid1).df_fixed;
    auto fixed2 = d_eatoms.at(aid2).df_fixed;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1971,26 +1841,23 @@ void EmbeddedFrag::removeCollisionsShortenBonds() {
         if (deg1 == 1) {
-          RDGeom::Point2D loc = d_eatoms[aid1].loc;
-          int aidA = _findDeg1Neighbor(dp_mol, aid1);
+          auto loc = d_eatoms[aid1].loc;
```
```suggestion
          auto loc = d_eatoms.at(aid1).loc;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -1971,26 +1841,23 @@ void EmbeddedFrag::removeCollisionsShortenBonds() {
         if (deg2 == 1 && !fixed2) {
-          RDGeom::Point2D loc = d_eatoms[aid2].loc;
-          int aidA = _findDeg1Neighbor(dp_mol, aid2);
+          auto loc = d_eatoms[aid2].loc;
```
```suggestion
          auto loc = d_eatoms.at(aid2).loc;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -2011,23 +1878,21 @@ void EmbeddedFrag::removeCollisionsShortenBonds() {
-        for (rpi = rPath.begin(); rpi != rPath.end(); rpi++) {
-          if (d_eatoms[*rpi].df_fixed) {
+        for (auto rpi : rPath) {
+          if (d_eatoms[rpi].df_fixed) {
```
```suggestion
          if (d_eatoms.at(rpi).df_fixed) {
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -2011,23 +1878,21 @@ void EmbeddedFrag::removeCollisionsShortenBonds() {
-          move = d_eatoms[nbrMap[*rpi][0]].loc;
-          move += d_eatoms[nbrMap[*rpi][1]].loc;
+          auto move = d_eatoms[nbrMap[rpi][0]].loc;
+          move += d_eatoms[nbrMap[rpi][1]].loc;
```
```suggestion
          move += d_eatoms.at(nbrMap.at(rpi).at(1)).loc;
```

greglandrum replied:
```suggestion
          move += d_eatoms[nbrMap.at(rpi)[1]].loc;
```

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -2011,23 +1878,21 @@ void EmbeddedFrag::removeCollisionsShortenBonds() {
+          move += d_eatoms[nbrMap[rpi][1]].loc;
           move *= 0.5;
-          move -= d_eatoms[*rpi].loc;
+          move -= d_eatoms[rpi].loc;
```
```suggestion
          move -= d_eatoms.at(rpi).loc;
```

ptosco commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -44,19 +44,18 @@ void embedFusedSystems(const RDKit::ROMol &mol,
-    for (RDKit::INT_VECT_CI rid = fused.begin(); rid != fused.end(); ++rid) {
-      frings.push_back(arings[*rid]);
+    for (auto rid : fused) {
+      frings.push_back(arings[rid]);
```
```suggestion
      frings.push_back(arings.at(rid));
```

ptosco commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -231,11 +227,10 @@ void computeInitialCoords(RDKit::ROMol &mol,
+      auto mrank = static_cast<int>(RDKit::MAX_INT);
+      RDKit::INT_LIST_I mnri;
+      for (auto nri = nratms.begin(); nri != nratms.end(); ++nri) {
+        auto rank = atomRanks[*nri];
```
```suggestion
        auto rank = atomRanks.at(*nri);
```

greglandrum commented:
> I hope I haven't done too many mistakes with my revision suggestions.

I'm looking through these now. I'm curious about all of the suggestions to switch from `d_eatoms[key]`->`d_eatoms.at(key)` when `d_eatoms` isn't const. Why do that?

greglandrum commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -737,6 +690,17 @@ void EmbeddedFrag::addAtomToAtomWithAng(unsigned int aid, unsigned int toAid) {
+    rtrans.TransformPoint(currLoc2);
+    if (findNumNeigh(currLoc, 0.5) > findNumNeigh(currLoc2, 0.5)) {
+      currLoc = currLoc2;
+      currAngle *= -1;
```
why is this better?

ptosco replied:
My thinking was that it's a 1-bit change as opposed to a multiplication. I realize that the compiler might end up generating identical code, or anyway make almost no difference. If you think it's worthless, just reject it.

greglandrum replied:
I find the `*=` form to be more immediately understandable, so I'll leave it as is.

ptosco commented:
@greglandrum
> I'm looking through these now. I'm curious about all of the suggestions to switch from d_eatoms[key]->d_eatoms.at(key) when d_eatoms isn't const. Why do that?

My reasoning was that (hopefully, if I understood the code right in my quick skimming) where I switched to `std::map::at` it was cases where a new item is not supposed to be added to the map, but it should rather be a read-only access. Thus if the key does not exist that would result in an exception, highlighting a possible bug in the code (or maybe bad input), rather than in a default value being created at that key. Again, if you think those are wrong and will rase undue exceptions because a new value is actually supposed to be created, just reject them. In a couple of cases I replaced a `CHECK_INVARIANT` `std::map::find` followed by access with `std::map::operator[]` with `std::map:;at`, and that does seem equivalent and more efficient/concise to me.

greglandrum commented:
> @greglandrum
> 
> > I'm looking through these now. I'm curious about all of the suggestions to switch from d_eatoms[key]->d_eatoms.at(key) when d_eatoms isn't const. Why do that?
> 
> My reasoning was that (hopefully, if I understood the code right in my quick skimming) where I switched to `std::map::at` it was cases where a new item is not supposed to be added to the map, but it should rather be a read-only access. Thus if the key does not exist that would result in an exception, highlighting a possible bug in the code (or maybe bad input), rather than in a default value being created at that key. Again, if you think those are wrong and will rase undue exceptions because a new value is actually supposed to be created, just reject them. In a couple of cases I replaced a `CHECK_INVARIANT` `std::map::find` followed by access with `std::map::operator[]` with `std::map:;at`, and that does seem equivalent and more efficient/concise to me.

That makes perfect sense. Thanks for the clarification!

greglandrum commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -489,35 +464,30 @@ RDGeom::Transform2D EmbeddedFrag::computeOneAtomTrans(
 RDGeom::Transform2D EmbeddedFrag::computeTwoAtomTrans(
     unsigned int aid1, unsigned int aid2,
     const RDGeom::INT_POINT2D_MAP &nringCor) {
+  CHECK_INVARIANT(d_eatoms.find(aid1) != d_eatoms.end(), "");
+  CHECK_INVARIANT(d_eatoms.find(aid2) != d_eatoms.end(), "");
```
I'm going to leave this one in. It's redundant with the changes you suggest below, but it's also a form of documentation of expectations, so I think it's worth keeping around

ptosco commented on Code/GraphMol/Depictor/EmbeddedFrag.cpp:
```diff
@@ -489,35 +464,30 @@ RDGeom::Transform2D EmbeddedFrag::computeOneAtomTrans(
 RDGeom::Transform2D EmbeddedFrag::computeTwoAtomTrans(
     unsigned int aid1, unsigned int aid2,
     const RDGeom::INT_POINT2D_MAP &nringCor) {
+  CHECK_INVARIANT(d_eatoms.find(aid1) != d_eatoms.end(), "");
+  CHECK_INVARIANT(d_eatoms.find(aid2) != d_eatoms.end(), "");
```
```suggestion
  auto aid1It = d_eatoms.find(aid1);
  auto aid2It = d_eatoms.find(aid2);
  CHECK_INVARIANT(aid1It != d_eatoms.end(), "");
  CHECK_INVARIANT(aid2It != d_eatoms.end(), "");
```
